sdcm/loader.py

```diff
@@ -11,143 +11,115 @@
-"""
-
-        tmp_dir_exporter = tempfile.mkdtemp(prefix='cassandra-stress-tools')
-        tmp_path_exporter = os.path.join(tmp_dir_exporter, 'cassandra_stress_exporter')
-        tmp_path_remote = '/tmp/cassandra_stress_exporter'
-        system_path_remote = '/usr/bin/cassandra_stress_exporter'
-        with open(tmp_path_exporter, 'w') as tmp_cfg_prom:
-            tmp_cfg_prom.write(cassandra_stress_exporter)
-        try:
-            self.node.remoter.send_files(src=tmp_path_exporter, dst=tmp_path_remote)
-            self.node.remoter.run('sudo mv %s %s' %
-                                  (tmp_path_remote, system_path_remote))
-            self.node.remoter.run('sudo chmod +x %s' % system_path_remote)
-        finally:
-            shutil.rmtree(tmp_dir_exporter)
-
-    def install(self, node):
-        self.node = node
-        self.exporter_install()
-
-        collectd_setup = ScyllaCollectdSetup()
-        collectd_setup.install(node)
+import threading
+
+import concurrent.futures
+import select
+
+
+class CassandraStressExporter(object):
+    def __init__(self, instance_name, metrics, cs_operation, cs_log_filename, loader_idx, cpu_idx):
+        self.executor = concurrent.futures.ThreadPoolExecutor(1)
+        self.future = None
+        self._stop_event = threading.Event()
+        self.metrics = metrics
+        self.cs_operation = cs_operation
+        self.cs_log_filename = cs_log_filename
+        self.cs_metric = self.metrics.create_gauge('collectd_cassandra_stress_%s_gauge' % self.cs_operation,
+                                                   'Gauge for cassandra stress metrics',
+                                                   ['cassandra_stress_%s' % self.cs_operation, 'instance', 'loader_idx', 'cpu_idx', 'type'])
+        self.instance_name = instance_name
+        self.loader_idx = loader_idx
+        self.cpu_idx = cpu_idx
+
+    def __enter__(self):
+        self.start()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.stop()
+
+    def set_metric(self, name, value):
+        self.cs_metric.labels(0, self.instance_name, self.loader_idx, self.cpu_idx, name).set(value)
+
+    def start(self):
+        self.future = self.executor.submit(self.run)
+        return self.future
+
+    def stop(self):
+        self._stop_event.set()
+        self.clear_metrics()
+
+    def stopped(self):
+        return self._stop_event.is_set()
+
+    def clear_metrics(self):
+        if self.cs_metric:
+            for metric_name in ['ops', 'lat_mean', 'lat_med', 'lat_perc_95', 'lat_perc_99', 'lat_perc_999', 'lat_max', 'errors']:
+                self.set_metric(metric_name, 0.0)
+
+    def follow_file(self, filename):
+        with open(filename, 'r') as input:
+            while not self.stopped():
+                poller = select.poll()
+                poller.register(input, select.POLLIN)
+                line = None
+                if poller.poll(100):
+                    line = input.readline()
```

Comment:
FIX: potentially losing parts of the line.

Reply:
I've tested it, and since readline is being used, it would just wait in readline, we won't lose info.

the only possible problem might be when stopping would hang if write stopped in the middle of a line.